### ISSUE TITLE
Destructuring for declaration prevents entangled type information / type implications

### ISSUE THREAD

**Issue by ole-flavia** [microsoft/TypeScript#34667]
**TypeScript Version:**  3.8.0-dev.20191022

**Search Terms:** destructuring

**Code**

```ts
type X = {x: true, y: number[]};
type Y = {x: false, y: undefined};
type Z = X | Y;

function a(x: Z) {
  if (x.x) {
    x.y.forEach(console.log);
  }
}

function b({x, y}: Z) {
  if (x) {
    y.forEach(console.log);
 // ^ Object is possibly 'undefined'
  }
}
```

**Expected behavior:**
Function `b` should compile.

**Actual behavior:**
Typescript does not detect any relation between `x` and `y` variables in function `b`, so it does not figure out that `x === true => y is number[]`.

**Playground Link:** http://www.typescriptlang.org/play/?ts=3.8.0-dev.20191022&ssl=1&ssc=1&pln=17&pc=1#code/C4TwDgpgBAGlC8UDeAPAXFYAnArhANFCBgHY4C2ARhFgNoC6AvgNwBQokUAmgsulADMAhgBsAzgSIYcJACYQBASxIRZLduGgAtXnAA+3NqwEyAxsEUB7ElCEAKfloCUyVlCiKBUBwDoULpDd3KBQfEB8BSywAUSFTAAs7U2sxSxEIHxFLAHMnNndGVkLjMwtrKEo7VEIQRgxnV3dPb39G4PDImLjE5JJU9MycvLcAehGoAD0oAHlKACsIcw8xKDBLMTFFShEQKAByGXklFVk9oMLCoA

**Related Issues:** None

**Comment by jack-williams** [comment]
Duplicate of #32639, and tracked at  #12184.
